diff --git a/ports/esp32/esp32_common.cmake b/ports/esp32/esp32_common.cmake
index 565f6feec..abca80a72 100644
--- a/ports/esp32/esp32_common.cmake
+++ b/ports/esp32/esp32_common.cmake
@@ -11,6 +11,9 @@ endif()
 # Include core source components.
 include(${MICROPY_DIR}/py/py.cmake)
 
+# Re-include the tildagon cmake, so it can influence components
+include(${MICROPY_PORT_DIR}/boards/tildagon/mpconfigboard.cmake)
+
 # CMAKE_BUILD_EARLY_EXPANSION is set during the component-discovery phase of
 # `idf.py build`, so none of the extmod/usermod (and in reality, most of the
 # micropython) rules need to happen. Specifically, you cannot invoke add_library.
@@ -101,7 +104,7 @@ list(APPEND MICROPY_SOURCE_PORT
     machine_pin.c
     machine_touchpad.c
     machine_dac.c
-    machine_i2c.c
+    #machine_i2c.c
     network_common.c
     network_lan.c
     network_ppp.c
diff --git a/ports/esp32/mpconfigport.h b/ports/esp32/mpconfigport.h
index e36d12bc0..7fff5a767 100644
--- a/ports/esp32/mpconfigport.h
+++ b/ports/esp32/mpconfigport.h
@@ -12,6 +12,10 @@
 #include "driver/i2s_std.h"
 #include "esp_wifi_types.h"
 
+#ifndef USB_SERIAL_JTAG_PACKET_SZ_BYTES
+#define USB_SERIAL_JTAG_PACKET_SZ_BYTES (64)
+#endif
+
 #ifndef MICROPY_CONFIG_ROM_LEVEL
 #define MICROPY_CONFIG_ROM_LEVEL            (MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES)
 #endif
