diff --git a/ports/esp32/esp32_common.cmake b/ports/esp32/esp32_common.cmake
index e928fb439..fde5160a5 100644
--- a/ports/esp32/esp32_common.cmake
+++ b/ports/esp32/esp32_common.cmake
@@ -11,6 +11,9 @@ endif()
 # Include core source components.
 include(${MICROPY_DIR}/py/py.cmake)
 
+# Re-include the tildagon cmake, so it can influence components
+include(${MICROPY_PORT_DIR}/boards/tildagon/mpconfigboard.cmake)
+
 # CMAKE_BUILD_EARLY_EXPANSION is set during the component-discovery phase of
 # `idf.py build`, so none of the extmod/usermod (and in reality, most of the
 # micropython) rules need to happen. Specifically, you cannot invoke add_library.
@@ -69,7 +72,7 @@ list(APPEND MICROPY_SOURCE_PORT
     machine_pin.c
     machine_touchpad.c
     machine_dac.c
-    machine_i2c.c
+    #machine_i2c.c
     network_common.c
     network_lan.c
     network_ppp.c
