diff --git a/python-ecosys/requests/requests/__init__.py b/python-ecosys/requests/requests/__init__.py
index 7401029..e55840b 100644
--- a/python-ecosys/requests/requests/__init__.py
+++ b/python-ecosys/requests/requests/__init__.py
@@ -43,6 +43,7 @@ def request(
     auth=None,
     timeout=None,
     parse_headers=True,
+    allow_redirects=True,
 ):
     redirect = None  # redirection url, None means no redirection
     chunked_data = data and getattr(data, "__next__", None) and not getattr(data, "__len__", None)
@@ -159,7 +160,7 @@ def request(
         s.close()
         raise
 
-    if redirect:
+    if redirect and allow_redirects:
         s.close()
         if status in [301, 302, 303]:
             return request("GET", redirect, None, None, headers, stream)
