diff --git a/python-ecosys/requests/requests/__init__.py b/python-ecosys/requests/requests/__init__.py
index a9a1836..555d5d2 100644
--- a/python-ecosys/requests/requests/__init__.py
+++ b/python-ecosys/requests/requests/__init__.py
@@ -43,6 +43,7 @@ def request(
     auth=None,
     timeout=None,
     parse_headers=True,
+    allow_redirects=True,
 ):
     if headers is None:
         headers = {}
@@ -178,7 +179,7 @@ def request(
         s.close()
         raise
 
-    if redirect:
+    if redirect and allow_redirects:
         s.close()
         if status in [301, 302, 303]:
             return request("GET", redirect, None, None, headers, stream)
