diff --git a/python-ecosys/requests/requests/__init__.py b/python-ecosys/requests/requests/__init__.py
index 4ca7489..04481e3 100644
--- a/python-ecosys/requests/requests/__init__.py
+++ b/python-ecosys/requests/requests/__init__.py
@@ -43,6 +43,7 @@ def request(
     auth=None,
     timeout=None,
     parse_headers=True,
+    allow_redirects=True,
 ):
     if headers is None:
         headers = {}
@@ -180,7 +181,7 @@ def request(
         s.close()
         raise
 
-    if redirect:
+    if redirect and allow_redirects:
         s.close()
         # Use the host specified in the redirect URL, as it may not be the same as the original URL.
         headers.pop("Host", None)
