diff --git a/ports/esp32/Makefile b/ports/esp32/Makefile
index 7f31ea691..1eb783685 100644
--- a/ports/esp32/Makefile
+++ b/ports/esp32/Makefile
@@ -112,4 +112,4 @@ size-files:
 # This is done in a dedicated build directory as some CMake cache values are not
 # set correctly if not all submodules are loaded yet.
 submodules:
-	$(Q)IDF_COMPONENT_MANAGER=0 idf.py $(IDFPY_FLAGS) -B $(BUILD)/submodules -D UPDATE_SUBMODULES=1 reconfigure
+	$(Q)idf.py $(IDFPY_FLAGS) -B $(BUILD)/submodules -D UPDATE_SUBMODULES=1 reconfigure
diff --git a/ports/esp32/esp32_common.cmake b/ports/esp32/esp32_common.cmake
index dedef6d78..f1fbb5897 100644
--- a/ports/esp32/esp32_common.cmake
+++ b/ports/esp32/esp32_common.cmake
@@ -31,6 +31,9 @@ endif()
 # Include core source components.
 include(${MICROPY_DIR}/py/py.cmake)
 
+# Re-include the tildagon cmake, so it can influence components
+include(${MICROPY_PORT_DIR}/boards/tildagon/mpconfigboard.cmake)
+
 # CMAKE_BUILD_EARLY_EXPANSION is set during the component-discovery phase of
 # `idf.py build`, so none of the extmod/usermod (and in reality, most of the
 # micropython) rules need to happen. Specifically, you cannot invoke add_library.
@@ -122,7 +125,7 @@ list(APPEND MICROPY_SOURCE_PORT
     machine_pin.c
     machine_touchpad.c
     machine_dac.c
-    machine_i2c.c
+    #machine_i2c.c
     network_common.c
     network_lan.c
     network_ppp.c
diff --git a/ports/esp32/mpconfigport.h b/ports/esp32/mpconfigport.h
index b5b7d63a5..7cd35d36c 100644
--- a/ports/esp32/mpconfigport.h
+++ b/ports/esp32/mpconfigport.h
@@ -12,6 +12,10 @@
 #include "driver/i2s_std.h"
 #include "esp_wifi_types.h"
 
+#ifndef USB_SERIAL_JTAG_PACKET_SZ_BYTES
+#define USB_SERIAL_JTAG_PACKET_SZ_BYTES (64)
+#endif
+
 #ifndef MICROPY_CONFIG_ROM_LEVEL
 #define MICROPY_CONFIG_ROM_LEVEL            (MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES)
 #endif
