diff --git a/extmod/nimble/modbluetooth_nimble.c b/extmod/nimble/modbluetooth_nimble.c
index 6b000f7..29e4b1f 100644
--- a/extmod/nimble/modbluetooth_nimble.c
+++ b/extmod/nimble/modbluetooth_nimble.c
@@ -1943,12 +1943,18 @@ static int ble_secret_store_read(int obj_type, const union ble_store_key *key, u
             break;
         }
         case BLE_STORE_OBJ_TYPE_OUR_SEC: {
-            // <type=our,addr,*>
-            // Find our secret for this remote device.
-            assert(ble_addr_cmp(&key->sec.peer_addr, BLE_ADDR_ANY)); // Must have address.
-            assert(key->sec.idx == 0);
-            key_data = (const uint8_t *)&key->sec.peer_addr;
-            key_data_len = sizeof(ble_addr_t);
+            if (ble_addr_cmp(&key->sec.peer_addr, BLE_ADDR_ANY)) {
+                // <type=our,addr,*> (single)
+                // Find our secret for this specific remote device.
+                assert(key->sec.idx == 0);
+                key_data = (const uint8_t *)&key->sec.peer_addr;
+                key_data_len = sizeof(ble_addr_t);
+            } else {
+                // <type=our,*> (with index)
+                // Iterate all our secrets.
+                key_data = NULL;
+                key_data_len = 0;
+            }
             break;
         }
         case BLE_STORE_OBJ_TYPE_CCCD: {
